File: Kotlin/time-needed-to-inform-all-employees.kt
class Solution {
    fun numOfMinutes(n: Int, headID: Int, manager: IntArray, informTime: IntArray): Int {
        // create an array of lists   
        var hierarchy = Array<ArrayList<Int>>(n, {ArrayList<Int>()})

        // each bucket is a manager and the list is all his subordinates
        for ((i, m) in manager.withIndex()) {
            if (m < 0)
                continue
            hierarchy[m].add(i)
        }

        // starting from the head manager 
        // recursively sum the informTime for his subordinates        
        return calcTime(headID, hierarchy, informTime)
    }

    // add the max of the subordinate managers to the current manager time
    fun calcTime(id: Int, hierarchy: Array<ArrayList<Int>>, informTime: IntArray): Int {
        var maxTime = 0
        for (employee in hierarchy[id]) {
            val time = calcTime(employee, hierarchy, informTime)
            if (maxTime < time)
                maxTime = time
        }
        return informTime[id] + maxTime
    }
}

File: TypeScript/time-needed-to-inform-all-employees.ts
function numOfMinutes(n: number, headID: number, manager: number[], informTime: number[]): number {
    const hierarchy: number[][] = Array(n)
    for (let i = 0; i < n; i++) {
        hierarchy[i] = []
    }
    for (let i = 0; i < n; i++) {
        const m = manager[i]
        if (m !== -1) {
            hierarchy[m].push(i)
        }
    }
    const calcTime = (manager) => {
        let maxTime = 0
        for (const sub of hierarchy[manager]) {
            maxTime = Math.max(maxTime, calcTime(sub))
        }
        return maxTime + informTime[manager]
    }
    return calcTime(headID)
};